cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -102,6 +169,15 @@ Status IsNullExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
     bit_util::SetBitsTo(out_bitmap, out_span->offset, out_span->length, false);
   }
 
+  const auto t = arr.type->id();
+  if (t == Type::SPARSE_UNION) {
+    SetSparseUnionLogicalNullBits(arr, out_bitmap, out_span->offset);
+  } else if (t == Type::DENSE_UNION) {
+    SetDenseUnionLogicalNullBits(arr, out_bitmap, out_span->offset);
+  } else if (t == Type::RUN_END_ENCODED) {
+    SetREELogicalNullBits(arr, out_bitmap, out_span->offset);
+  }
+
```

Comment:
This whole block can be moved into the `else` above to make it more clear that this only gets called *after* a complete zeroing of `out_bitmap`.

Now we have to know about a logical implication that is not explicit in the code: `t \in {UNION, REE} \implies arr.GetNullCount() == 0`. That is what in turn guarantees `out_bitmap` is zeroed before the `Set*LogicalNullBits`. Too much logic. :sweat_smile:

Reply:
The `// Input has no nulls => output is entirely false.` also needs to be more nuanced after that move.